spec/services/display_ad_event_rollup_spec.rb

```diff
@@ -0,0 +1,116 @@
+require "rails_helper"
+
+RSpec.describe DisplayAdEventRollup, type: :service do
+  let(:ad1) { create :display_ad }
+  let(:ad2) { create :display_ad }
+  let(:user1) { create :user }
+  let(:user2) { create :user }
+
+  def override_timestamps
+    DisplayAdEvent.record_timestamps = false
+    yield
+    DisplayAdEvent.record_timestamps = true
+  end
+
+  def days_ago_as_range(num)
+    (Date.current - num.days).all_day
+  end
+
+  it "fails if new attributes would be lost" do
+    attributes_considered = DisplayAdEventRollup::ATTRIBUTES_PRESERVED + DisplayAdEventRollup::ATTRIBUTES_DESTROYED
+    expect(DisplayAdEvent.column_names.map(&:to_sym)).to contain_exactly(*attributes_considered)
+  end
```

Comment:
Love this!

just to be clear, this means that if there are any changes to the column, the aggregator will stop working until we figure out what to do?

Reply:
If a later migration adds a column on `display_ad_events`, the aggregator really _should not_ aggregate, until we put that column into either "preserve" (eg, group by) or "ignore". What this test accomplishes, is that the _test_ should **fail** if a new column is added to the `display_ad_events` table. That is, it shouldn't be possible for someone in the future to add a column without updating the rollup service.

It's hard to imagine what columns we might add in the future, but for examples of current columns: we "destroy" `updated_at` (the new, squashed row does not attempt to maintain the original rows' `updated_at` timestamps) but we "preserve" `category` ("clicks" are grouped separately from "impressions"). If a future column is more like `updated_at`, we can add it to the destroy list, if it's more like `category`, we should preserve it (eg, modify the rollup service to add it to the aggregator).